### PR TITLE
Ignore VersionNegotiation containing reserved values and the value of the used QUIC version

### LINKED ISSUE #675: We often fail to establish connection when handshake packet is corrupted
Test handshakecorruption fails very often:
[neqo-client.1.log](https://github.com/mozilla/neqo/files/4695444/neqo-client.1.log)
[neqo-client.2.log](https://github.com/mozilla/neqo/files/4695445/neqo-client.2.log)

### PR DESCRIPTION
**Pull request by ddragana**
fixes #675

This also add test server_receive_unknow_first_packet which was crashing when trying to get timers (I have accidentally made a wrong test that revealed the issue)

### CODE REVIEW

File: neqo-transport/tests/server.rs [hunk 1]
```diff
@@ -717,14 +717,10 @@ fn version_negotiation() {
     }
     assert!(found, "valid version not found");
 
+    // Client ignores VN packet that contain negotiatied version.
```
Comment (agrover) [inline neqo-transport/tests/server.rs:720]: typo: negotiated

File: neqo-transport/src/packet.rs [hunk 2]
```diff
@@ -614,6 +614,20 @@ impl<'a> PublicPacket<'a> {
             Err(Error::DecryptError)
         }
     }
+
+    pub fn decode_version_negotiation(&self) -> Res<Vec<Version>> {
```
Comment (agrover) [inline neqo-transport/src/packet.rs:618]: let's call this something a little less about the implementation details and more about the result. Maybe `supported_versions()`? Or you tell me.

File: neqo-transport/src/packet.rs [hunk 3]
```diff
@@ -614,6 +614,20 @@ impl<'a> PublicPacket<'a> {
...
     }
+
+    pub fn supported_versions(&self) -> Res<Vec<Version>> {
+        assert_eq!(self.packet_type, PacketType::VersionNegotiation);
+        let mut decoder = Decoder::new(&self.data[self.header_len..]);
+        let mut res = Vec::new();
+        while decoder.remaining() > 0 {
+            let version = Version::try_from(Self::opt(decoder.decode_uint(4))?)?;
+            // Ignore reserved versions.
+            if (version & 0x0f0f_0f0f) != 0x0a0a_0a0a {
```
Comment (martinthomson) [inline neqo-transport/src/packet.rs:625]: See above.

**Comment by ddragana** [comment]
> I'm a little uncomfortable with the idea that we might ignore reserved versions. It goes against the spirit of having reserved versions. I think that if we receive a response that contains only reserved versions, we should just let the connection fail.

you are right.

File: neqo-transport/src/connection.rs [hunk 4]
```diff
@@ -1026,6 +1026,15 @@ impl Connection {
                     self.crypto.states.init(self.role, &packet.dcid());
                 }
                 (PacketType::VersionNegotiation, State::WaitInitial, Role::Client) => {
+                    if let Ok(versions) = packet.supported_versions() {
+                        if versions.contains(&QUIC_VERSION) {
+                            // Ignore VersionNegotiation packets that contain the current version or only
+                            // contain reserved versions. (decode_version_negotiation will remove reserved
+                            // versions)
```
Comment (martinthomson) [inline neqo-transport/src/connection.rs:1033]: Update this comment.

```suggestion
                            // Ignore VersionNegotiation packets that contain the current version.
```

File: neqo-transport/src/connection.rs [hunk 5]
```diff
@@ -5173,4 +5183,141 @@ mod tests {
...
+    }
+
+    #[test]
+    fn version_negotiation_current_version() {
+        let mut client = default_client();
+
+        let vn = get_vn(&mut client, &[0x1a1_a1a1a, QUIC_VERSION]);
+
+        assert_eq!(
+            client.process(Some(Datagram::new(loopback(), loopback(), vn,)), now(),),
```
Comment (martinthomson) [inline neqo-transport/src/connection.rs:5240]: This is an odd time to be processing a VN packet.  Can you make this a little more conventional in the sense that it is received after sending the first Initial?

File: neqo-transport/src/connection.rs [hunk 6]
```diff
@@ -5173,4 +5183,141 @@ mod tests {
...
+            client.process(
+                Some(Datagram::new(loopback(), loopback(), &vn[..vn.len() - 1])),
+                now(),
+            ),
+            Output::None
+        );
+        match client.state() {
+            State::Closed(err) => {
+                assert_eq!(*err, ConnectionError::Transport(Error::VersionNegotiation))
+            }
```
Comment (martinthomson) [inline neqo-transport/src/connection.rs:5281]: Shouldn't we be ignoring corrupted version negotiation packets too?

File: neqo-transport/src/connection.rs [hunk 7]
```diff
@@ -5173,4 +5183,141 @@ mod tests {
...
+
+    #[test]
+    fn version_negotiation_empty() {
+        let mut client = default_client();
+
+        let vn = get_vn(&mut client, &[]);
+
+        assert_eq!(
+            client.process(
+                Some(Datagram::new(loopback(), loopback(), &vn[..vn.len() - 1])),
```
Comment (martinthomson) [inline neqo-transport/src/connection.rs:5294]: Check the truncation here.

File: neqo-transport/src/connection.rs [hunk 8]
```diff
@@ -5173,4 +5183,141 @@ mod tests {
...
+        match client.state() {
+            State::Closed(err) => {
+                assert_eq!(*err, ConnectionError::Transport(Error::VersionNegotiation))
+            }
+            _ => panic!("Invalid client state"),
+        }
+    }
+
+    #[test]
+    fn version_negotiation_empty() {
```
Comment (martinthomson) [inline neqo-transport/src/connection.rs:5287]: Shouldn't this have the same effect as the corrupted one?  That is, shouldn't we ignore it?

File: neqo-transport/src/connection.rs (hunk 5 above)
Comment (ddragana) [reply neqo-transport/src/connection.rs:5240]: The initial is sent inside get_vn(). I moved it there because we need the Intitial packet to retrieve cids.

**Comment by agrover** [comment]
@ddragana please resolve conflicts.
@martinthomson please re-review
I'd like to get this in asap.

File: neqo-transport/src/connection.rs (hunk 5 above)
Comment (martinthomson) [reply neqo-transport/src/connection.rs:5244]: OK, that's misunderstanding on my part.  Renaming get_vn to handle_vn might work, but you would have to have it return the result of the client.process() call. The advantage of that is that you save another line of code from every test.

File: neqo-transport/src/connection.rs [hunk 9]
```diff
@@ -5173,4 +5187,131 @@ mod tests {
...
+        assert_eq!(
+            server.process(
+                Some(Datagram::new(
+                    loopback(),
+                    loopback(),
+                    unknown_version_packet,
+                )),
+                now(),
+            ),
+            Output::None
```
Comment (martinthomson) [inline neqo-transport/src/connection.rs:5207]: Ahh, the reason that this is not generating a VN response is that we generate that in Server not Connection.  That took me longer to realize than I should be happy to admit.  Could you add a comment explaining that?

**Pull request merged**